### PR TITLE
adding support for a passing the json config in string form to CodeModificationStep

### PR DESCRIPTION
**Pull request by deepchoudhery**
- split `CodeChangeHelper` into its individual classes
  - removed `CodeChangeOptions` and just using a List<string> to track these options. Will be reworked and made simpler in another upcoming PR.
- make most of the classes in `Microsoft.DotNet.Scaffolding.CodeModification` internal, only `CodeModificationStep` is public and that is by design.
- modified `ProjectModifierHelper.FilterCodeSnippets` to not use `CodeChangeOptions` and use a List<string>
- modified `CodeModificationStep` to handle both `CodeModifierConfigPath` (we load this text, deserialize the json into a `CodeModifierConfig` object) or `CodeModifierConfigJsonText` (deserialize the json text from this string instead)
  - prioritizing `CodeModifierConfigJsonText` over `CodeModifierConfigPath` (will make it clearer in the documentation)
- added `GlobalToolFileFinder`: moved logic from `CodeModifierConfigHelper` to here for finding 'CodeModificationConfigs' folder in the tool for these json configs. 
- changed `CodeSnippet` (now inaccessible for 1st and 3rd party tools) to json string values for the same object. 
- created a new `CodeModificationStep` to add the above changes in `BlazorCrudScaffolderBuilderExtensions.cs`

### CODE REVIEW

File: src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.Internal/GlobalToolFileFinder.cs [hunk 1]
```diff
@@ -0,0 +1,47 @@
+using System.Diagnostics;
+using System.Reflection;
+
+namespace Microsoft.DotNet.Scaffolding.Internal;
+
+internal static class GlobalToolFileFinder
+{
+    internal static string? FindCodeModificationConfigFile(string fileName, Assembly executingAssembly)
+    {
+        var assemblyDirectory = Path.GetDirectoryName(executingAssembly.Location);
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.Internal/GlobalToolFileFinder.cs:10]: i'll add a null check yes ty

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Extensions/ScaffolderBuilderExtensions.cs [hunk 2]
```diff
@@ -31,15 +32,19 @@ internal static class ScaffolderBuilderExtensions
                 return;
             }
 
+            string? connectionStringName = null;
             context.Properties.TryGetValue("BaseProjectPath", out var baseProjectPathObj);
+            context.Properties.TryGetValue("CodeModifierProperties", out var codeModifierPropertiesObj);
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/dotnet-scaffold-aspnet/Extensions/ScaffolderBuilderExtensions.cs:37]: yep I really should, I was rushing this PR, will do that

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs [hunk 3]
```diff
@@ -32,78 +31,12 @@ internal static class BlazorCrudHelper
...
-        {
-            { "Create", CreateBlazorTemplate },
-            { "Delete", DeleteBlazorTemplate },
-            { "Details", DetailsBlazorTemplate },
-            { "Edit", EditBlazorTemplate },
-            { "Index", IndexBlazorTemplate }
-        });
-
-    internal static Dictionary<string, string> CRUDTemplates => _crudTemplates.Value;
+    internal static string AdditionaCodeModificationJson = "{ \"Files\": [ { \"FileName\": \"Program.cs\", \"Options\": [], \"Methods\": { \"Global\": { \"CodeChanges\": [$(CodeChanges)] } } } ] }";
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs:34]: will fix ty

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs [hunk 4]
```diff
@@ -32,78 +31,12 @@ internal static class BlazorCrudHelper
...
+    internal static string AddRazorComponentsSnippet = "{\"Block\":\"WebApplication.CreateBuilder.Services.AddRazorComponents()\",\"InsertBefore\":[\"var app = WebApplication.CreateBuilder.Build()\"],\"CodeChangeType\":\"Default\",\"LeadingTrivia\":{\"Newline\":true,\"NumberOfSpaces\":0}}";
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs:35]: yep that would be better, switching to that

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Extensions/ScaffolderBuilderExtensions.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/dotnet-scaffold-aspnet/Extensions/ScaffolderBuilderExtensions.cs:37]: fixed ty

File: src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.CodeModification/CodeModificationStep.cs [hunk 5]
```diff
@@ -24,29 +32,49 @@ public CodeModificationStep(ILogger<CodeModificationStep> logger)
...
+            codeModifierConfig = CodeModifierConfigHelper.GetCodeModifierConfig(CodeModifierConfigPath);
+        }
+        else
+        {
+            return false;
+        }
+        
+        if (codeModifierConfig is null)
+        {
+            _logger.LogError($"Unable to parse the {nameof(CodeModifierConfig)} provided. Please check the {nameof(CodeModifierConfig)} definition.");
```
[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.CodeModification/CodeModificationStep.cs:51]: sure yeah I need better error messaging here anyways

[the remark this answers is not in the document]
Comment (deepchoudhery) [reply src/dotnet-scaffolding/Microsoft.DotNet.Scaffolding.CodeModification/CodeModificationStep.cs:51]: added some placeholder error messaging, will go over that with the PMs and adjust as needed.

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs [hunk 6]
```diff
@@ -31,12 +31,81 @@ internal static class BlazorCrudHelper
...
-    internal static string AddMapRazorComponentsSnippet = "{\"Block\":\"app.MapRazorComponents<App>()\",\"InsertBefore\":[\"app.Run()\"],\"CodeChangeType\":\"Default\",\"LeadingTrivia\":{\"Newline\":true,\"NumberOfSpaces\":0}}";
-    internal static string AddInteractiveServerRenderModeSnippet = "{\"Block\":\"AddInteractiveServerRenderMode()\",\"Parent\":\"MapRazorComponents<App>\",\"CodeChangeType\":\"MemberAccess\",\"LeadingTrivia\":{\"Newline\":true,\"NumberOfSpaces\":4}}";
-    internal static string AddInteractiveServerComponentsSnippet = "{\"Block\":\"AddInteractiveServerComponents()\",\"Parent\":\"WebApplication.CreateBuilder.Services.AddRazorComponents()\",\"CodeChangeType\":\"MemberAccess\",\"LeadingTrivia\":{\"Newline\":true,\"NumberOfSpaces\":4}}";
-    internal static string AddInteractiveWebAssemblyRenderModeSnippet = "{\"Block\":\"AddInteractiveWebAssemblyRenderMode()\",\"Parent\":\"MapRazorComponents<App>\",\"CodeChangeType\":\"MemberAccess\",\"LeadingTrivia\":{\"Newline\":true,\"NumberOfSpaces\":4}}";
+    internal static string AdditionalCodeModificationJson = @"
```
Comment (anvillan) [inline src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs:34]: Make all the new entries constant values => `internal const string`.

**Review by anvillan — APPROVED** [review]

File: src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs (hunk 6 above)
Comment (deepchoudhery) [reply src/dotnet-scaffolding/dotnet-scaffold-aspnet/Helpers/BlazorCrudHelper.cs:34]: oops yesyes

**Pull request merged**